server/etcdserver/server.go

```diff
@@ -1643,6 +1644,10 @@ func (s *EtcdServer) AppliedIndex() uint64 { return s.getAppliedIndex() }
 
 func (s *EtcdServer) Term() uint64 { return s.getTerm() }
 
+// ProbeRaftLoopProgress probes if the etcdserver raft loop is deadlocked
+// since last time the function is invoked.
+func (s *EtcdServer) ProbeRaftLoopProgress() bool { return s.r.safeReadTickElapsedAndClear() != 0 }
```

Comment:
Probe response should not be determined based on frequency of probing. If someone probes etcd more frequently then proposals are happening then this will return failure.

I still need to think more about the best approach (maybe a document that collects and compares the approaches would help?), but this could be improved by waiting for a tick instead of failing if there was no ticks since last probe.

Reply:
> Probe response should not be determined based on frequency of probing. If someone probes etcd more frequently then proposals are happening then this will return failure.

Thanks, in that case maybe adding a simple throttler in etcdserver raft Node should help. 

> I still need to think more about the best approach (maybe a document that collects and compares the approaches would help?), but this could be improved by waiting for a tick instead of failing if there was no ticks since last probe.

Yeah, a small doc compares with pros and cons should help.

Reply:
> I am also interested in how apiserver implements /livez and /readyz.

https://kubernetes.io/docs/reference/using-api/health-checks/

https://github.com/kubernetes/kubernetes/pull/70676#discussion_r232921395

@siyuanfoundation @logicalhan

Reply:
As far as I can tell, health probes in apiserver generally are not complicated. They mostly checks if something has started or initialized, such as https://github.com/kubernetes/kubernetes/blob/c7d270302c8de3afc9d7b01c70faf3a18407ce44/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L173

Reply:
Addeded throttler to make raft loop prober check effective at most once every 5 seconds.

Reply:
As promised, I wrote a small doc to compare the pros and cons. 

Could you please take a look? @ahrtr @serathius @siyuanfoundation Thanks~

Reply:
Thanks. Overall this PR looks good to me, also added comments in the doc.